File: 2021/day10/day10.go
package main

import (
	"fmt"
	"io/ioutil"
	"sort"
	"strings"
)

func get_data() []string {
	data, _ := ioutil.ReadFile("input.txt")
	data = data[:len(data)-1]
	lines := strings.Split(string(data), "\n")
	return lines
}

func main() {
	left := map[string]bool{"(": true, "[": true, "{": true, "<": true}
	right := map[string]string{")": "(", "]": "[", "}": "{", ">": "<"}
	errors := map[string]int{")": 3, "]": 57, "}": 1197, ">": 25137}
	complete := map[string]int{"(": 1, "[": 2, "{": 3, "<": 4}

	lines := get_data()

	error_score := 0             // part 1
	completion_scores := []int{} // part2

	for _, line := range lines {
		stack := []string{}
		for _, ru := range line {
			char := string(ru)
			if _, found := left[char]; found {
				stack = append(stack, char)
			} else {
				matching := right[char]
				if stack[len(stack)-1] != matching {
					error_score += errors[char]
					stack = []string{}
					break
				} else {
					stack = stack[:len(stack)-1]
				}
			}
		}

		// part 2
		if len(stack) > 0 {
			complete_score := 0
			for i := len(stack) - 1; i >= 0; i-- {
				complete_score *= 5
				complete_score += complete[stack[i]]
			}
			completion_scores = append(completion_scores, complete_score)
		}
	}

	fmt.Println(error_score)

	sort.Ints(completion_scores)
	fmt.Println(completion_scores[len(completion_scores)/2])
}

File: 2021/day10/day10.pl
#!/usr/bin/perl
use strict;
use warnings;

my %error = (')' => 3, ']' => 57, '}' => 1197, '>' => 25137);
my %complete = ('(' => 1, '[' => 2, '{' => 3, '<' => 4);

my $fh;
open $fh, '<', 'input.txt';
chomp(my @lines = <$fh>);
close $fh;

my $error_score = 0;
my @complete_scores;
for my $line (@lines) {
	my $len = 0;
	while ($len != length($line)) {
		$len = length($line);
		$line =~ s/\(\)|\[\]|\{\}|\<\>//g;
	}
	my $li = $line;
	$line =~ s/^[\(\[\{\<]+//;
	if (length($line) > 0) {
		$error_score += $error{substr($line, 0, 1)};
	}
	else {
		$li = reverse($li);
		my $complete_score = 0;
		$li =~ s/(.)/$complete_score = $complete_score * 5 + $complete{$1}/ge;
		push @complete_scores, $complete_score;
	}	
}

print "$error_score\n";
@complete_scores = sort { $a <=> $b } @complete_scores;
print $complete_scores[@complete_scores / 2] . "\n";
